### PR TITLE
[Enhancement] Check if key exists when inserting keys to LakePersistentIndex

### PR DESCRIPTION
**Pull request by TszKitLo40**
## Why I'm doing:
Keys are not checked whether they exist in sstables. If they are not checked, duplicated key can not be found.
## What I'm doing:
1. Check if key exists when inserting keys to LakePersistentIndex
3. Rename `_block_cache` to `_pk_index_block_cache` in `update_manager`.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/lake/lake_persistent_index.cpp [hunk 1]
```diff
@@ -205,10 +210,16 @@ Status LakePersistentIndex::upsert(size_t n, const Slice* keys, const IndexValue
 
 Status LakePersistentIndex::insert(size_t n, const Slice* keys, const IndexValue* values, int64_t version) {
     RETURN_IF_ERROR(_memtable->insert(n, keys, values, version));
+    KeyIndexSet key_indexes;
+    for (int i = 0; i < n; ++i) {
+        key_indexes.insert(n);
+    }
+    for (auto iter = _sstables.rbegin(); iter != _sstables.rend(); ++iter) {
```
Comment (wyb) [inline be/src/storage/lake/lake_persistent_index.cpp:217]: also check immutable memtable?

Comment (TszKitLo40) [reply be/src/storage/lake/lake_persistent_index.cpp:220]: added

File: be/src/storage/lake/lake_persistent_index.cpp [hunk 2]
```diff
@@ -205,10 +210,19 @@ Status LakePersistentIndex::upsert(size_t n, const Slice* keys, const IndexValue
 
 Status LakePersistentIndex::insert(size_t n, const Slice* keys, const IndexValue* values, int64_t version) {
     RETURN_IF_ERROR(_memtable->insert(n, keys, values, version));
+    KeyIndexSet key_indexes;
+    for (int i = 0; i < n; ++i) {
+        key_indexes.insert(n);
+    }
+    if (_immutable_memtable != nullptr) {
+        RETURN_IF_ERROR(_immutable_memtable->check_not_exist(keys, key_indexes, version));
```
Comment (kevincai) [inline be/src/storage/lake/lake_persistent_index.cpp:218]: what's the meaning of return error after the `_memtable->insert` is done? shouldn't check the existence first and then do the insertion?

Comment (TszKitLo40) [reply be/src/storage/lake/lake_persistent_index.cpp:218]: If duplicated key is found in `_memtable`, `_memtable->insert` will return error. So if `_memtable->insert` returns success, only `immutable_memtable` and `ssts` are needed to be checked.

Comment (kevincai) [reply be/src/storage/lake/lake_persistent_index.cpp:218]: what's the issue this PR try to fix?

let's say key A are not in _memtable, but in immutable_memtable. what is the expected behavior when doing insert(A)? will it report error or success, will the _memtable be polluted even the error reported to caller?

**Review by kevincai — COMMENTED** [review]
unit test should be covered to the `insert()` interface as well, will the insert report correctly when duplicate key exists in memtable/immutable_memtable/sstable, should check if there is any modification happens if insert fails.

File: be/src/storage/lake/lake_persistent_index.cpp (hunk 2 above)
Comment (TszKitLo40) [reply be/src/storage/lake/lake_persistent_index.cpp:218]: In current implementation, `insert` is only used in reload. If error is reported, persistent index is in an error state. The persistent index can not be used any more.

**Comment by TszKitLo40** [comment]
> unit test should be covered to the `insert()` interface as well, will the insert report correctly when duplicate key exists in memtable/immutable_memtable/sstable, should check if there is any modification happens if insert fails.

done

**Review by wyb — DISMISSED** [review]

File: be/src/storage/lake/lake_persistent_index.cpp (hunk 2 above)
Comment (TszKitLo40) [reply be/src/storage/lake/lake_persistent_index.cpp:219]: done

**Review by kevincai — DISMISSED** [review]

**Review by kevincai — DISMISSED** [review]
code conflict.

**Comment by TszKitLo40** [comment]
Duplicated keys can not be checked easily now. For example, in version 3, key "aaa" is erased but it is not flushed to sstable.  When index is recovered, when "aaa" is checked, we will find that there is a "aaa" whose version is 2. So duplicated keys will be found. This PR will be closed now.

**Pull request closed without merging**